### ISSUE TITLE
Merging with transformers library -- Random results

### ISSUE THREAD

**Issue by shtoshni92** [facebookresearch/SpanBERT#14]
I just wanted to check if you are planning on merging this with the transformers library since that would make it immensely easier to use. For now, just extracting the archive and calling the Bert APIs has worked for me but the solution feels inelegant as I have to download the model separately. Surprisingly, loading the BertModel using this codebase and the tokenizer using the transformers library results in random results across runs (with the same batch order and content). Can't really point out where the issue is for now!

**Comment by mandarjoshi90** [comment]
Hi Shubham,
Merging with transformers isn't top priority for now. I'm not sure I understand the second part of the post. Are you finetuning SpanBERT and seeing variance across seeds? If so, how much?

**Comment by shtoshni92** [comment]
Hi Mandar,
So I am not finetuning and just using BertModel.from_pretrained('spanbert-base-cased'). Because the current URL points to .tar.gz, I have to use your codebase which is based on pytorch_pretrained_bert (using the tar URL with current transformers API makes it thinks it's a TF checkpoint.)
But for tokenization, I'm using the BertTokenizer from transformers API just because it provides a sort of unified interface for many models + the SpanBERT model uses the same Tokenizer from what I could understand. 
For some reason, the combination leads to results that vary across runs (no finetuning involved!).  

BTW I understand that merging with transformers is not a priority but just to clarify, do you think for non-finetuning results it's just a matter of calling BERT based APIs with SpanBERT model's initialization. Also, for finetuning what might change?

**Comment by shtoshni92** [comment]
Ah sorry, I figured out the issue. So transformers initializes the pretrained model to be in eval() mode by default but that's not the case with SpanBERT codebase. The randomization in dropout causes random results :)

**Comment by phosseini** [comment]
> Ah sorry, I figured out the issue. So transformers initializes the pretrained model to be in eval() mode by default but that's not the case with SpanBERT codebase. The randomization in dropout causes random results :)

Did you manage to convert the code to a transformer friendly version? I am trying to convert the run_*.py codes to a transformer-based format, but I'm having some troubles: [https://github.com/facebookresearch/SpanBERT/issues/48#issue-652709306](url)

**Comment by shtoshni92** [comment]
No I just used the SpanBERT codebase to load the pretrained model. 
Regarding your issue, I think the model rarely results just logits. Typically it's a list, tuple, or I think in the latest codebase they made it a class - see [here](https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_bert.py#L807). I think you should inspect the model output, check it's type etc.

**Comment by phosseini** [comment]
> No I just used the SpanBERT codebase to load the pretrained model.
> Regarding your issue, I think the model rarely results just logits. Typically it's a list, tuple, or I think in the latest codebase they made it a class - see [here](https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_bert.py#L807). I think you should inspect the model output, check it's type etc.

Thanks for the feedback. I finally managed to write transformers compatible version - see [here](https://github.com/facebookresearch/SpanBERT/issues/48#issuecomment-666897648) it needs some more testing though for TACRED, but I could run it on my relation extraction task/data which follows the same format as TACRED.
